docs/source/developers/contributing.rst

```diff
@@ -124,29 +181,72 @@ To contribute a patch:
+  for maintainers to accept.
 * Add new unit tests for your code.
+* Follow the style guides for the part(s) of the project you're modifying.
+  Some languages (C++, Python, and Rust, for example) run a lint check in
+  continuous integration. For all languages, see their respective developer
+  documentation and READMEs for style guidance. In general, try to make it look
+  as if the codebase has a single author, and emulate any conventions you see,
+  whether or not they are officially documented or checked.
+
+When tests are passing and the pull request has been approved by the interested
+parties, a committer will merge the pull request. This is done with a
+command-line utility that does a squash merge, so all of your commits will be
+registered as a single commit to the master branch; this simplifies the
+connection between JIRA issues and commits, and it makes it easier to bisect
+history to identify where changes were introduced. A side effect of this way of
+merging is that your pull request will appear in the GitHub interface to have
+been "closed without merge". Do not be alarmed: if you look at the bottom, you
+will see a message that says "@user closed this in $COMMIT".
+
+Local git conventions
++++++++++++++++++++++
+
+If you are tracking the Arrow source repository locally, here are some tips
+for using ``git``.
+
+All Arrow contributors work off of their personal fork of ``apache/arrow``
+and submit pull requests "upstream". Once you've cloned your fork of Arrow,
+be sure to::
+
+    $ git remote add upstream https://github.com/apache/arrow
+
+to set the "upstream" repository.
+
+You are encouraged to develop on branches, rather than your own "master" branch,
+and it helps to keep your fork's master branch synced with ``upstream/master``.
 
-Thank you in advance for your contributions!
+To start a new branch, pull the latest from upstream first::
 
-Common Git conventions followed within the project
---------------------------------------------------
+   $ git fetch upstream
+   $ git checkout master
+   $ git reset --hard upstream/master
+   $ git checkout -b $NEW_BRANCH_NAME
 
-If you are tracking the Arrow source repository locally, following some common Git
-conventions would make everyone's workflow compatible.  These recommendations along with
-their rationale are outlined below.
+It does not matter what you call your branch. Some people like to use the JIRA
+number as branch name, others use descriptive names.
 
-It is strongly discouraged to use a regular ``git merge``, as a linear commit history is
-prefered by the project.  It is much easier to maintain, and makes for easier
-``cherry-picking`` of features; useful for backporting fixes to maintenance releases.
+Once you have a branch going, you should sync with ``upstream/master``
+regularly, as many commits merge to master every day.
+It is recommended to use ``git rebase`` rather than ``git merge``.
```

Comment:
Should this "recommended" be a bit stronger? 

As I think quite some people are used to a merge-workflow rather than a rebase-workflow (including myself), and my understanding was that we really prefer the rebasing workflow (for reasons that I don't understand ;)).

Reply:
Resolving conflicts in a PR with merge commits can be a nightmare, whereas resolving the same conflicts in a PR that doesn't have the merges is, to me, relatively straightforward. Maybe I'm doing something wrong. Very rarely have I seen anything good come out of `git merge`

So I think we can say "The project maintainers prefer to handle conflict resolution from a rebase standpoint rather than from a merge standpoint. If a PR contains merge commits and contains conflicts, then a maintainer may squash the changes and rebase them to remove the merge commits."

Reply:
FWIW at large organizations like Google, PRs (which are called "changelists") with more than one commit are not accepted. Every proposed patch must be a single rebased commit (this is tightly integrated with Gerrit, Google's code review system). I would be much happier using Gerrit than GitHub PRs because of the incremental code diffing system -- once you experience it, it is difficult to go back.

Reply:
> Resolving conflicts in a PR with merge commits can be a nightmare

If you only `merge upstream/master`, then that is in my experience never a nightmare (you should of course never mix rebasing and merging, as that indeed will give nightmares). 

Also, when merging master instead of rebasing, you only need to fix merge conflicts once. While now we have here a complicated section about how to simplify the conflict resolution by squashing your commits while rebasing. This is never needed in a merging workflow (which IMO is also nicer because it preserves that you can see what has been changed since a last review on github).

Anyway, we are not going to resolve that discussion here ;) (and both workflows have its pros/cons). I was mainly wondering to what extent we want to push contributors to a chosen workflow.

Reply:
Yes, it's partly a lack of GitHub's reviewing functionalities, but *when* using GitHub (which is the case right now), my feeling is that a merging workflow works better with it.

Reply:
Honestly I don't think it matters what people do on their branches since we squash merge in the end, right?

The last thing I want out of this document is to spark a flame war between `rebase` and `merge` partisans. I wouldn't have added this section myself but there was a JIRA requesting it and someone else added it this weekend.